### ISSUE TITLE
Wireguard PreShared Key doesn't work in LuCI

### ISSUE THREAD

**Issue by lleachii** [openwrt/luci#1798]
Steps to reproduce:

* Create a Wireguard interface
* Add and establish a peer connection
* Add a preshared key to both client and LuCI device
* Reboot the machine running LuCI to apply the new config (as hitting Save/Apply and restarting networking does not work)

The devices continue to handshake, but pass no traffic.

* On the machine with LuCI, running the wg command at the command prompt does not display the added Pre-shared Key for the Wireguard interface.

**Comment by jow-** [comment]
Does LuCI produce a correct configuration?

**Comment by lleachii** [comment]
Yes, it does create an appropriate line in /etc/config/network:

`        option preshared_key 'XXXXXXXXXXXXXXXXXXXXXXXXXXXXXX='`

**Comment by jow-** [comment]
I have no idea at all. I never used wireguard and am not even sure how to setup a test environment. Also as far as LuCI is concerned, it seems to generate a correct configuration.

As for it taking three days; in the meanwhile I worked on migrating the mailing lists, preparing a new OpenWrt release, answered support questions on forum, irc and mailing lists, debugged firewall issues and problems with the LuCI HTTP parser library, asked people versed with wireguard to check for the psk issue. Oh, and I had to attend my day job as well. So *please* be patient. If you think it is so simple then take a look at the wireguard script yourself: https://github.com/openwrt/openwrt/blob/master/package/network/services/wireguard/files/wireguard.sh

**Comment by lleachii** [comment]
Jow, thanks.

I reviewed that SH file and this: https://github.com/openwrt/luci/blob/master/protocols/luci-proto-wireguard/luasrc/model/cbi/admin_network/proto_wireguard.lua (which you first question suggests isnt the issue, because the config file is proprtly made.

Also, the link you provide is for the creation of the CONF file. As I noted the PresharedKey properly appears in the file.

I haven't located anything odd. I wonder if **wg** was compiled without that feature. I'm still looking, thanks for pointing me in the right direction!

**Comment by lleachii** [comment]
Jow,

I figured it out! The PresharedKey is implemented **in LuCI as a per Interface option. This is invalid.** The preshared key should be a **per Peer** option.

It SH file appears that it would correctly parse this; but it's not properly implemented in UCI/LuCI?

**Comment by lleachii** [comment]
> "wireguard_" .. ifname, suggesting that it is actually about interfaces.

I always wondered why the developer named the peer sections for the WG interfaces in that manner.

But yes, PresharedKey should appear under the individual peer section as applicable; but it's actually created under the master WG interface, which is invalid.

**Comment by dibdot** [comment]
@lleachii could you please post a working/tested wg config for reference?

**Comment by lleachii** [comment]
Here you go. This is a Interface and Peer config:
```
# interface config
config interface 'test_wireguard'                                                                                                                                        
        option proto 'wireguard'                                                                                                                                         
        option private_key 'XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX='                                                                                                
        option listen_port '45698'                                                                                                                                       
        list addresses '172.44.44.1/29'

# peer config
config wireguard_test_wireguard                                                                                                                                          
        option public_key 'YYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYY='                                                                                                 
        list allowed_ips '172.44.44.2/32' 
```

The PSK should appear under in the Peer section; but actually appears in the Interface section.

**Comment by zx2c4** [comment]
Nice observation. Indeed the preshared key is an attribute of the peer not the interface, so it needs to be set on the peer.

> It actually looks like the whole "peers" section is actually a combination of peers and interfaces options, as the peers section is named with "wireguard_" .. ifname, suggesting that it is actually about interfaces.

No, the peers section is only information about peers, not about the interface; it's not mixed. However, I agree that this whole business of putting peer information in a top level thing prefixed with "wireguard_" is awkward and strange. I'll defer to @danrl to explain his reasoning for that and determine whether it should be changed.

**Comment by dibdot** [comment]
@lleachii Sorry, I can't reproduce your issue with current master. I've created a (dummy) wireguard interface and added two peers ... the PSK will be added to the peers ... that's the generated part of /etc/config/network:

```
config interface 'test_wireguard'
        option proto 'wireguard'
        option listen_port '5555'
        list addresses '172.44.44.1'
        option private_key 'XXXXXXXXXXqea6KDN8l3JcF5XbfawHoKyrkT6DN2BlE='

config wireguard_test_wireguard
        option preshared_key 'XXXXXXXXXXqea6KDN8l3JcF5XbfawHoKyrkT6DN2BlE='
        option public_key 'XXXXXXXXXXqea6KDN8l3JcF5XbfawHoKyrkT6DN2BlE='
        list allowed_ips '192.168.1.1'

config wireguard_test_wireguard
        option public_key XXXXXXXXXXqea6KDN8l3JcF5XbfawHoKyrkT6DN2BlE='
        option preshared_key 'XXXXXXXXXXqea6KDN8l3JcF5XbfawHoKyrkT6DN2BlE='
        list allowed_ips '192.168.1.2'

```

**Comment by lleachii** [comment]
That's impossible, as the LuCI app doesn't have a PresharedKey section the Peer area, it's under the main Interface.

> luci-app-wireguard | git-18.118.20919-cfba791-1

**Comment by dibdot** [comment]
It' an optional peer option ...
![peer](https://user-images.githubusercontent.com/8061346/40256346-6efc0952-5aea-11e8-8a1a-bdecbc3a0dcc.png)

**Comment by lleachii** [comment]
Oh my...I stand corrected...what version of OpenWRT are you running?

I'm running 17.01.4 and as I noted, I'm running luci-app-wireguard | git-18.118.20919-cfba791-1.

That **'additional field'** section does not appear on my device. I see PresharedKey on the **Advanced** tab directly below MTU.

**Comment by dibdot** [comment]
as I said before, I used current master ... please retest with that version.

root@adb2go:~# opkg list-installed | grep wireguard
kmod-wireguard - 4.9.100+0.0.20180514-1
luci-app-wireguard - git-18.138.59467-72fe5dd-1
luci-proto-wireguard - git-18.138.59467-72fe5dd-1
wireguard-tools - 0.0.20180514-1

**Comment by lleachii** [comment]
I must admit, I never had a clue what "master" was. This is my most extensive bug report to an OSS development project.

> please retest with that version.

Are you referring to **"Snapshot"**?

**Comment by dibdot** [comment]
"master" is the main development tree, makes always sense to test against that ... especially with early birds like wireguard. You can setup your own buildroot environment or use a snapshot release.

**Comment by dibdot** [comment]
... and I think you can close this ticket ... ;-)

**Comment by lleachii** [comment]
@dibdot ,

Thanks...I learned a lot more today about how to locate the most up-to-date versions of the packages. A previous bug I reported was fixed in 17.01.4, I just figured the same would occur as fixes were made. This means everyone on 17.01.4 will continue to have this issue. I'll have to better follow how version are marked to understand this.

Thanks again all.

**Comment by danrl** [comment]
Yep, I moved it to a per peer option when the change in WireGuard was made. Please alsways use master as long as this is experimental :) Thanks.
